ipatests/test_integration/test_acme.py

```diff
@@ -0,0 +1,236 @@
+            [
+                'certbot',
+                '--server', self.acme_server,
+                'certonly',
+                '--domain', self.clients[0].hostname,
+                '--standalone',
+            ],
+        )
+
+    @pytest.mark.skipif(skip_certbot_tests, reason='certbot not available')
+    def test_certbot_revoke(self):
+        # Assume previous certonly operation succeeded.
+        # Read certificate to learn serial number.
+        cert_path = \
+            f'/etc/letsencrypt/live/{self.clients[0].hostname}/cert.pem'
+        data = self.clients[0].get_file_contents(cert_path)
+        cert = x509.load_pem_x509_certificate(data, backend=default_backend())
+
+        # revoke cert via ACME
+        self.clients[0].run_command(
+            [
+                'certbot',
+                '--server', self.acme_server,
+                'revoke',
+                '--cert-name', self.clients[0].hostname,
+                '--delete-after-revoke',
+            ],
+        )
+
+        # check cert is revoked (kinit already performed)
+        result = self.master.run_command(
+            ['ipa', 'cert-show', str(cert.serial_number), '--raw']
+        )
+        assert 'revocation_reason:' in result.stdout_text
+
+    @pytest.mark.skipif(skip_certbot_tests, reason='certbot not available')
+    def test_certbot_dns(self):
+        # Assume previous revoke operation succeeded and cert was deleted.
+        # We can now request a new certificate.
+
+        # Get a cert from ACME service using dns-01 challenge and Certbot's
+        # standalone HTTP server mode
+        self.clients[0].run_command([
+            'certbot',
+            '--server', self.acme_server,
+            'certonly',
+            '--non-interactive',
+            '--domain', self.clients[0].hostname,
+            '--preferred-challenges', 'dns',
+            '--manual',
+            '--manual-public-ip-logging-ok',
+            '--manual-auth-hook', CERTBOT_DNS_IPA_SCRIPT,
+            '--manual-cleanup-hook', CERTBOT_DNS_IPA_SCRIPT,
+        ])
+
+    ##############
+    # mod_md tests
+    ##############
+
+    @pytest.mark.skipif(skip_mod_md_tests, reason='mod_md not available')
+    def test_mod_md(self):
+        # write config
+        self.clients[0].run_command(['mkdir', '-p', '/etc/httpd/conf.d'])
+        self.clients[0].put_file_contents(
+            '/etc/httpd/conf.d/md.conf',
+            '\n'.join([
+                f'MDCertificateAuthority {self.acme_server}',
+                'MDCertificateAgreement accepted',
+                f'MDomain {self.clients[0].hostname}',
+                '<VirtualHost *:443>',
+                f'    ServerName {self.clients[0].hostname}',
+                '    SSLEngine on',
+                '</VirtualHost>\n',
+            ]),
+        )
+
+        # restart and wait a while, then graceful restart to load cert
+        self.clients[0].run_command(['systemctl', 'restart', 'httpd'])
+        time.sleep(15)
```

Comment:
Can you loop waiting for the cert to get issued instead of the sleep?

Reply:
To do that we would have to couple tightly to mod_md's "database" format (i.e. how it stores the certs it retrieves), or to the log messages it produces.  I'm not keen to do either because the tests will break if those things change.  But I can add a log message explaning this, and why I prefer a conservative sleep.

Reply:
I should elaborate that if mod_md did not require the subsequent graceful restart to begin using the cert, then a loop would be a reasonable approach.

Reply:
I've updated the commit to add this commentary and it will arrive in the next push.